src/server/saved_objects/migrations/kibana/kibana_migrator.ts

```diff
@@ -0,0 +1,172 @@
+      (path: 'migrations.batchSize' | 'migrations.pollInterval'): number;
+    };
+  };
+  plugins: { elasticsearch: ElasticsearchPlugin | undefined };
+}
+
+interface ElasticsearchPlugin {
+  getCluster: ((name: 'admin') => { callWithInternalUser: CallCluster });
+  waitUntilReady: () => Promise<any>;
+}
+
+/**
+ * Manages the shape of mappings and documents in the Kibana index.
+ *
+ * @export
+ * @class KibanaMigrator
+ */
+export class KibanaMigrator {
+  private kbnServer: KbnServer;
+  private documentMigrator: VersionedTransformer;
+  private mappingProperties: MappingProperties;
+
+  /**
+   * Creates an instance of KibanaMigrator.
+   *
+   * @param opts
+   * @prop {KbnServer} kbnServer - An instance of the Kibana server object.
+   * @memberof KibanaMigrator
+   */
+  constructor({ kbnServer }: { kbnServer: KbnServer }) {
+    this.kbnServer = kbnServer;
+    this.mappingProperties = mergeProperties(kbnServer.uiExports.savedObjectMappings || []);
+    this.documentMigrator = new DocumentMigrator({
+      kibanaVersion: kbnServer.version,
+      migrations: kbnServer.uiExports.savedObjectMigrations || {},
+      validateDoc: docValidator(kbnServer.uiExports.savedObjectValidations || {}),
+    });
+  }
+
+  /**
+   * Gets the index mappings defined by Kibana's enabled plugins.
+   *
+   * @returns
+   * @memberof KibanaMigrator
+   */
+  public getActiveMappings() {
+    return buildActiveMappings({ properties: this.mappingProperties });
+  }
+
+  /**
+   * Migrates an individual doc to the latest version, as defined by the plugin migrations.
+   *
+   * @param {SavedObjectDoc} doc
+   * @returns {SavedObjectDoc}
+   * @memberof KibanaMigrator
+   */
+  public migrateDocument(doc: SavedObjectDoc): SavedObjectDoc {
+    return this.documentMigrator.migrate(doc);
+  }
+
+  /**
+   * Migrates the mappings and documents in the Kibana index.
+   *
+   * @returns
+   * @memberof KibanaMigrator
+   */
+  public migrateIndex() {
+    const { server } = this.kbnServer;
+
+    // We can't do anything if the elasticsearch plugin has been disabled.
+    if (!server.plugins.elasticsearch) {
+      server.log(
+        ['warning', 'migration'],
+        'The elasticsearch plugin is disabled. Skipping migrations.'
+      );
+      return { status: 'skipped' };
+    }
+
+    const config = server.config();
+    const migrator = new IndexMigrator({
+      batchSize: config.get('migrations.batchSize'),
+      callCluster: createCallCluster(server),
+      documentMigrator: this.documentMigrator,
+      index: config.get('kibana.index'),
+      log: (meta: string[], message: string) => server.log(meta, message),
+      mappingProperties: this.mappingProperties,
+      pollInterval: config.get('migrations.pollInterval'),
+      scrollDuration: config.get('migrations.scrollDuration'),
+    });
+
+    return migrator.migrate();
+  }
+}
+
+/**
+ * Wait until the elasticsearch plugin says it's ready, then return the
```

Comment:
question: This is primarily necessary because we don't have strong guarantees around the lifecycle of our elasticsearch functionality in the legacy world, right?

Reply:
Yes, presumably this might go away with the new platform. The need for this can be tested by starting Kibana without elasticsearch running. Then, after a while, boot up elastic search. Migrations kick in once Kibana connects to elasticsearch.